### PR TITLE
[Optimization](statistics)  optimize Incremental statistics collection and statistics cleaning

### PR DESCRIPTION
**Pull request by weizhengte**
# Proposed changes

This pr mainly optimizes the following items:
- the collection of statistics: clear up invalid historical statistics before collecting them, so as not to affect the final table statistics.
- the incremental collection of statistics: in the case of incremental collection, only the corresponding partition statistics need to be collected.

TODO: Supports incremental collection of materialized view statistics.

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
run p0

**Comment by weizhengte** [comment]
run clickbench

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.57 seconds
 stream load tsv:          445 seconds loaded 74807831229 Bytes, about 160 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230414011929_clickbench_pr_128828.html

File: fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisHelper.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+
+/**
+ * Used to help collect statistics.
+ * TODO add other method
+ */
+public class AnalysisHelper extends MasterDaemon {
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisHelper.java:35]: This class name is a little confused, in convention we name a class as XXXHelper when it's stateless and just provide some procedural code

File: fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisHelper.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    private final ThreadPoolExecutor executors = ThreadPoolManager.newDaemonThreadPool(
+            Config.statistics_helper_running_task_num,
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisHelper.java:38]: Rename this config option either

File: fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisHelper.java (hunk 1 above)
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisHelper.java:35]: I thnk there is no need to add master daemon for this, you'd better simply mark let analyzemanager to do this

File: fe/fe-common/src/main/java/org/apache/doris/common/Config.java [hunk 3]
```diff
@@ -1695,6 +1695,24 @@ public class Config extends ConfigBase {
...
+     * if true, statistics will be updated automatically
+     */
+    @ConfField(mutable = true, masterOnly = true)
+    public static boolean enable_auto_collect_statistics = false;
+
+    /*
+     * collect statistics periodically can be stressful on the cluster
+     */
+    @ConfField(mutable = true, masterOnly = true)
+    public static boolean enable_period_collect_statistics = false;
```
Comment (Kikyou1997) [inline fe/fe-common/src/main/java/org/apache/doris/common/Config.java:1708]: What's diff between enable_period_collect_statistics and enable_auto_collect_statistics, besides these two options isn't used anywhere

File: fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisHelper.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            Config.statistics_helper_running_task_num, 0,
+            TimeUnit.DAYS, new LinkedBlockingQueue<>(),
+            new BlockedPolicy("AnalysisHelper Executor", Integer.MAX_VALUE),
+            "AnalysisHelper Executor", true);
+
+    public Future<?> asyncExecute(Runnable runnable) {
+        return executors.submit(runnable);
+    }
+}
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisHelper.java:47]: Seems this class is only partially implemented....it even didn't overwrite runAfterCatalogReady, what's the meaning of commit this unrelated code in this PR?

File: fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsRepository.java [hunk 5]
```diff
@@ -179,15 +205,7 @@ private static void dropStatistics(Long dbId,
...
-            String statement = isHistogram ? stringSubstitutor.replace(DROP_TABLE_HISTOGRAM_TEMPLATE) :
-                    stringSubstitutor.replace(DROP_TABLE_STATISTICS_TEMPLATE);
-            StatisticsUtil.execUpdate(statement);
-        } catch (Exception e) {
-            LOG.warn("Drop statistics failed", e);
-        }
+        return params;
     }
 
     private static <T> void buildPredicate(String fieldName, Set<T> fieldValues, StringBuilder predicate) {
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsRepository.java:211]: generic type doesn't make any sense

File: fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisManager.java [hunk 6]
```diff
@@ -100,16 +98,24 @@ public void createAnalysisJob(AnalyzeStmt analyzeStmt) throws DdlException {
...
                     .collect(Collectors.toSet());
-            StatisticsRepository.dropStatistics(dbId, tblIds, colNames, partIds);
+            try {
+                StatisticsRepository.markStatsAsDeleted(dbId, tblIds, colNames, partIds);
+            } catch (Exception e) {
+                throw new DdlException("Fail to mark statistics as deleted", e);
+            }
+            // Removing statistics maybe be slow and use asynchronous method
+            Env.getCurrentEnv().getAnalysisHelper().asyncExecute(() ->
+                    StatisticsRepository.dropStatistics(dbId, tblIds, colNames, partIds));
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisManager.java:118]: It's really puzzling here, if you has an insatiable craving for deletion here AND submit a async task to do this, then what's the necessary to update `column_statistics` table synchronously to mark a row should be deleted instead of let async task delete them directly

**Review by Kikyou1997 — CHANGES_REQUESTED** [review]
Further modifications and discussions are needed for this PR before it could be merged

**Comment by weizhengte** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsRepository.java (hunk 5 above)
Comment (weizhengte) [reply fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsRepository.java:253]: fieldValues could be Set<Long> tblIds, or Set<String> colNames... all converted in this method

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
> Further modifications and discussions are needed for this PR before it could be merged

I have removed the code that is not part of this pr~

File: fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsRepository.java (hunk 5 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsRepository.java:254]: you're right, thx

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
run p0

**Comment by weizhengte** [comment]
run clickbench

**Review by Kikyou1997 — DISMISSED** [review]

**Comment by weizhengte** [comment]
Wait until the sampling statistics pr are comerged before modifying

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
run clickbench

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
run clickbench
